File: src/packages/control-ui/src/api/pac_subscription_orchestrator.ts
import { HealthTier, SubscriptionHealthClassifier } from './subscription_health_classifier.js';
import { NodePoolAggregator } from './node_pool_aggregator.js';
import { SubscriptionCrawlerPipeline } from './subscription_crawler_pipeline.js';
import { PacDiffSynchronizer } from './pac_diff_synchronizer.js';
import type { PacSyncDelta } from './pac_diff_synchronizer.js';
import { PacRuleAction, PacRuleGenerator } from './pac_rule_generator.js';

export interface OrchestratorSummary {
  totalCrawled: number;
  totalActivePool: number;
  usableTierACount: number;
  currentPacRulesCount: number;
  pacChecksum: string;
}

export class PacSubscriptionOrchestrator {
  private crawler: SubscriptionCrawlerPipeline;
  private pool: NodePoolAggregator;
  private classifier: SubscriptionHealthClassifier;
  private pacGen: PacRuleGenerator;
  private pacSync: PacDiffSynchronizer;
  public defaultProxyPort: number;

  constructor(initialDomains: string[] = [], defaultProxyPort: number = 10808) {
    this.defaultProxyPort = defaultProxyPort;
    this.pacGen = new PacRuleGenerator(PacRuleAction.Direct);
    for (const d of initialDomains) {
      this.pacGen.addRule(d, PacRuleAction.Proxy, `127.0.0.1:${defaultProxyPort}`);
    }

    this.crawler = new SubscriptionCrawlerPipeline();
    this.pool = new NodePoolAggregator();
    this.classifier = new SubscriptionHealthClassifier(10);
    this.pacSync = new PacDiffSynchronizer(initialDomains);
  }

  registerSubscriptionSource(url: string, intervalSecs: number = 3600): void {
    this.crawler.addSource(url, intervalSecs);
  }

  executeCrawlAndIngest(sourceUrl: string, rawContent: string, now: number): number {
    const count = this.crawler.ingestCrawlContent(sourceUrl, rawContent);
    const proxies = this.crawler.getHarvestedProxies();
    this.pool.ingestRawEntries(proxies, now);
    return count;
  }

  recordNodeProbe(nodeId: string, latencyMs: number, success: boolean): void {
    this.classifier.recordSample(nodeId, latencyMs, success);
    this.pool.updateHealth(nodeId, latencyMs, success);
  }

  updatePacWithUpstream(upstreamDomains: string[]): PacSyncDelta {
    const delta = this.pacSync.computeDelta(upstreamDomains);
    this.pacSync.applyDelta(delta);

    // Rebuild pacGen
    this.pacGen = new PacRuleGenerator(PacRuleAction.Direct);
    for (const domain of upstreamDomains) {
      this.pacGen.addRule(domain, PacRuleAction.Proxy, `127.0.0.1:${this.defaultProxyPort}`);
    }

    return delta;
  }

  exportActivePacScript(): string {
    return this.pacGen.generatePacScript();
  }

  getSummary(): OrchestratorSummary {
    const tierA = this.classifier.filterUsableNodes(HealthTier.TierAExcellent);
    return {
      totalCrawled: this.crawler.totalHarvestedCount(),
      totalActivePool: this.pool.rankNodes(0.0).length,
      usableTierACount: tierA.length,
      currentPacRulesCount: this.pacSync.totalRules(),
      pacChecksum: this.pacSync.currentChecksum(),
    };
  }
}

File: labs/mobile/donor-archive/tunnel/PacSubscriptionOrchestrator.kt
package com.luminet.android.tunnel

data class AndroidOrchestratorSummary(
    val totalCrawled: Int,
    val totalActivePool: Int,
    val usableTierACount: Int,
    val currentPacRulesCount: Int,
    val pacChecksum: String
)

class PacSubscriptionOrchestrator(initialDomains: List<String>, val defaultProxyPort: Int = 10808) {
    private val crawler = SubscriptionCrawlerPipeline()
    private val pool = NodePoolAggregator()
    private val classifier = SubscriptionHealthClassifier(10)
    private var pacGen = PacRuleGenerator(PacRuleAction.DIRECT)
    private val pacSync = PacDiffSynchronizer(initialDomains)

    init {
        for (d in initialDomains) {
            pacGen.addRule(d, PacRuleAction.PROXY, "127.0.0.1:$defaultProxyPort")
        }
    }

    fun registerSubscriptionSource(url: String, intervalSecs: Long) {
        crawler.addSource(url, intervalSecs)
    }

    fun executeCrawlAndIngest(sourceUrl: String, rawContent: String, now: Long): Int {
        val count = crawler.ingestCrawlContent(sourceUrl, rawContent)
        val proxies = crawler.getHarvestedProxies()
        pool.ingestRawEntries(proxies, now)
        return count
    }

    fun recordNodeProbe(nodeId: String, latencyMs: Int, success: Boolean) {
        classifier.recordSample(nodeId, latencyMs, success)
        pool.updateHealth(nodeId, latencyMs, success)
    }

    fun updatePacWithUpstream(upstreamDomains: List<String>): PacSyncDelta {
        val delta = pacSync.computeDelta(upstreamDomains)
        pacSync.applyDelta(delta)

        pacGen = PacRuleGenerator(PacRuleAction.DIRECT)
        for (d in upstreamDomains) {
            pacGen.addRule(d, PacRuleAction.PROXY, "127.0.0.1:$defaultProxyPort")
        }

        return delta
    }

    fun exportActivePacScript(): String {
        return pacGen.generatePacScript()
    }

    fun getSummary(): AndroidOrchestratorSummary {
        val tierA = classifier.filterUsableNodes(HealthTier.TIER_A_EXCELLENT)
        return AndroidOrchestratorSummary(
            totalCrawled = crawler.totalHarvestedCount(),
            totalActivePool = pool.rankNodes(0.0).size,
            usableTierACount = tierA.size,
            currentPacRulesCount = pacSync.totalRules(),
            pacChecksum = pacSync.currentChecksum()
        )
    }
}
